File: convex/lib/frontmatter.ts
/**
 * Parse YAML frontmatter from a markdown string.
 * Returns { frontmatter, body }.
 *
 * Supports arbitrary nesting depth via recursive descent:
 *   metadata:
 *     strawpot:
 *       dependencies:
 *         skills:
 *           - git-workflow
 *         roles:
 *           - reviewer
 *   → { metadata: { strawpot: { dependencies: { skills: ["git-workflow"], roles: ["reviewer"] } } } }
 */
export function parseFrontmatter(text: string): {
  frontmatter: Record<string, unknown>;
  body: string;
} {
  const match = text.match(/^---\s*\n([\s\S]*?\n)---\s*\n([\s\S]*)$/);
  if (!match) {
    return { frontmatter: {}, body: text };
  }

  const yamlStr = match[1];
  const body = match[2];
  const lines = yamlStr.split("\n");

  const { result } = parseBlock(lines, 0, 0);
  return { frontmatter: result, body };
}

function getIndent(line: string): number {
  return line.search(/\S/);
}

function skipBlanks(lines: string[], from: number): number {
  let i = from;
  while (i < lines.length && !lines[i].trim()) i++;
  return i;
}

function parseScalar(raw: string): unknown {
  // Quoted string
  if (
    (raw.startsWith('"') && raw.endsWith('"')) ||
    (raw.startsWith("'") && raw.endsWith("'"))
  ) {
    return raw.slice(1, -1);
  }

  // Boolean
  if (raw === "true") return true;
  if (raw === "false") return false;

  // Number
  if (/^\d+(\.\d+)?$/.test(raw)) {
    return parseFloat(raw);
  }

  // Bare string
  return raw;
}

function parseInlineValue(raw: string): unknown {
  // Inline array: [a, b, c]
  if (raw.startsWith("[") && raw.endsWith("]")) {
    return raw
      .slice(1, -1)
      .split(",")
      .map((s) => s.trim().replace(/^["']|["']$/g, ""))
      .filter(Boolean);
  }

  return parseScalar(raw);
}

function parseBlock(
  lines: string[],
  startIdx: number,
  baseIndent: number,
): { result: Record<string, unknown>; nextIdx: number } {
  const result: Record<string, unknown> = {};
  let i = startIdx;

  while (i < lines.length) {
    const line = lines[i];
    const trimmed = line.trim();

    if (!trimmed) {
      i++;
      continue;
    }

    const indent = getIndent(line);
    if (indent < baseIndent) break;

    const kvMatch = trimmed.match(/^(\w[\w-]*):\s*(.*)$/);
    if (!kvMatch) {
      i++;
      continue;
    }

    const [, key, rawValue] = kvMatch;
    const value = rawValue.trim();
    i++;

    if (value) {
      result[key] = parseInlineValue(value);
      continue;
    }

    // Empty value — peek to determine array vs nested object
    const peekIdx = skipBlanks(lines, i);

    if (peekIdx >= lines.length) {
      result[key] = "";
      continue;
    }

    const peekIndent = getIndent(lines[peekIdx]);
    if (peekIndent <= indent) {
      result[key] = "";
      continue;
    }

    const peekTrimmed = lines[peekIdx].trim();

    if (peekTrimmed.startsWith("- ")) {
      const parsed = parseArray(lines, peekIdx, peekIndent);
      result[key] = parsed.result;
      i = parsed.nextIdx;
    } else if (peekTrimmed.match(/^(\w[\w-]*):\s*(.*)$/)) {
      const parsed = parseBlock(lines, peekIdx, peekIndent);
      result[key] = parsed.result;
      i = parsed.nextIdx;
    } else {
      i = peekIdx + 1;
    }
  }

  return { result, nextIdx: i };
}

function parseArray(
  lines: string[],
  startIdx: number,
  baseIndent: number,
): { result: string[]; nextIdx: number } {
  const result: string[] = [];
  let i = startIdx;

  while (i < lines.length) {
    const line = lines[i];
    const trimmed = line.trim();

    if (!trimmed) {
      i++;
      continue;
    }

    const indent = getIndent(line);
    if (indent < baseIndent) break;
    if (!trimmed.startsWith("- ")) break;

    result.push(trimmed.slice(2).trim());
    i++;
  }

  return { result, nextIdx: i };
}

/**
 * Extract dependencies from parsed frontmatter.
 *
 * Reads from metadata.strawpot.dependencies.
 * For skills: returns { skills: string[] } from a flat array.
 * For roles: returns { skills?: string[], roles?: string[] } from a nested object.
 */
export function extractDependencies(
  fm: Record<string, unknown>,
  kind: "skill" | "role",
): { skills?: string[]; roles?: string[] } | undefined {
  const meta = fm.metadata as Record<string, unknown> | undefined;
  const strawpot = meta?.strawpot as Record<string, unknown> | undefined;
  const deps = strawpot?.dependencies;

  if (deps == null) return undefined;

  if (kind === "skill" && Array.isArray(deps)) {
    return { skills: deps as string[] };
  }
  if (kind === "role" && typeof deps === "object" && !Array.isArray(deps)) {
    return deps as { skills?: string[]; roles?: string[] };
  }

  return undefined;
}

File: cli/src/strawhub/frontmatter.py
"""Parse YAML frontmatter from markdown strings.

Supports arbitrary nesting depth via recursive descent:
  metadata:
    strawpot:
      dependencies:
        skills:
          - git-workflow
        roles:
          - reviewer
  → {"metadata": {"strawpot": {"dependencies": {"skills": ["git-workflow"], "roles": ["reviewer"]}}}}

Python port of convex/lib/frontmatter.ts.
"""

import re


def parse_frontmatter(text: str) -> dict:
    """Parse YAML frontmatter from a markdown string.

    Returns {"frontmatter": {...}, "body": "..."}.
    """
    m = re.match(r"^---\s*\n([\s\S]*?\n)---\s*\n([\s\S]*)$", text)
    if not m:
        return {"frontmatter": {}, "body": text}

    yaml_str = m.group(1)
    body = m.group(2)
    lines = yaml_str.split("\n")

    result, _ = _parse_block(lines, 0, 0)
    return {"frontmatter": result, "body": body}


def _get_indent(line: str) -> int:
    return len(line) - len(line.lstrip())


def _skip_blanks(lines: list[str], start: int) -> int:
    i = start
    while i < len(lines) and not lines[i].strip():
        i += 1
    return i


def _parse_scalar(raw: str):
    """Parse a scalar value: quoted string, boolean, number, or bare string."""
    # Quoted string
    if (raw.startswith('"') and raw.endswith('"')) or (
        raw.startswith("'") and raw.endswith("'")
    ):
        return raw[1:-1]

    # Boolean
    if raw == "true":
        return True
    if raw == "false":
        return False

    # Number
    if re.match(r"^\d+(\.\d+)?$", raw):
        return float(raw) if "." in raw else int(raw)

    # Bare string
    return raw


def _parse_inline_value(raw: str):
    """Parse an inline value (scalar or inline array)."""
    if raw.startswith("[") and raw.endswith("]"):
        return [
            s.strip().strip("\"'")
            for s in raw[1:-1].split(",")
            if s.strip()
        ]
    return _parse_scalar(raw)


def _parse_block(
    lines: list[str], start_idx: int, base_indent: int
) -> tuple[dict, int]:
    """Recursively parse a YAML block at a given indent level.

    Returns (result_dict, next_index).
    """
    result: dict = {}
    i = start_idx

    while i < len(lines):
        line = lines[i]
        trimmed = line.strip()

        if not trimmed:
            i += 1
            continue

        indent = _get_indent(line)
        if indent < base_indent:
            break

        kv = re.match(r"^(\w[\w-]*):\s*(.*)$", trimmed)
        if not kv:
            i += 1
            continue

        key = kv.group(1)
        value = kv.group(2).strip()
        i += 1

        if value:
            result[key] = _parse_inline_value(value)
            continue

        # Empty value — peek to determine array vs nested object
        peek_idx = _skip_blanks(lines, i)

        if peek_idx >= len(lines):
            result[key] = ""
            continue

        peek_indent = _get_indent(lines[peek_idx])
        if peek_indent <= indent:
            result[key] = ""
            continue

        peek_trimmed = lines[peek_idx].strip()

        if peek_trimmed.startswith("- "):
            arr, next_idx = _parse_array(lines, peek_idx, peek_indent)
            result[key] = arr
            i = next_idx
        elif re.match(r"^(\w[\w-]*):\s*(.*)$", peek_trimmed):
            nested, next_idx = _parse_block(lines, peek_idx, peek_indent)
            result[key] = nested
            i = next_idx
        else:
            i = peek_idx + 1

    return result, i


def _parse_array(
    lines: list[str], start_idx: int, base_indent: int
) -> tuple[list[str], int]:
    """Parse consecutive '- item' lines at a given indent level.

    Returns (result_list, next_index).
    """
    result: list[str] = []
    i = start_idx

    while i < len(lines):
        line = lines[i]
        trimmed = line.strip()

        if not trimmed:
            i += 1
            continue

        indent = _get_indent(line)
        if indent < base_indent:
            break
        if not trimmed.startswith("- "):
            break

        result.append(trimmed[2:].strip())
        i += 1

    return result, i


def extract_dependencies(
    fm: dict, kind: str
) -> dict | None:
    """Extract dependencies from parsed frontmatter.

    Reads from metadata.strawpot.dependencies.
    For skills: returns {"skills": [...]} from a flat array.
    For roles: returns {"skills": [...], "roles": [...]} from a nested object.
    """
    deps = (
        fm.get("metadata", {})
        .get("strawpot", {})
        .get("dependencies")
    )

    if deps is None:
        return None

    if kind == "skill" and isinstance(deps, list):
        return {"skills": deps}
    if kind == "role" and isinstance(deps, dict):
        return deps

    return None
